dashboard/test/factories/factories.rb

```diff
@@ -422,6 +422,34 @@
     login_type 'email'
 
     initialize_with {Section.new(attributes)}
+
+    factory :mega_section do
+      after(:create) do |mega_section|
+        200.times do
+          student = create :student
+          mega_section.students << student
+        end
+        script = create :script
+        mega_section.script = script
+        i = 0
+        100.times do
+          create(
+            :script_level,
+            script: script,
+            levels: [
+              create(:maze, name: "test level #{i}")
+            ]
+          )
+          i += 1
+        end
+        mega_section.students.each do |student|
+          script.script_levels.each do |script_level|
+            create :user_level, user: student, level: script_level.levels.first, script: script_level.script
+          end
```

Comment:
I'm pretty sure the slow down is in the these nested loops.  The [benchmark](https://ruby-doc.org/stdlib-2.5.0/libdoc/benchmark/rdoc/Benchmark.html) for the steps in creating a mega_section up until the loops is ~40 seconds; with the loops ~350 seconds. 😱

Reply:
I am asking FactoryBot to make 20,000 user_levels so... no wonder it takes a long time.

Reply:
200 students * 100 script levels, so 20,000 items written to the db one at a time? that would probably do it...

(Can you remind me whether this is actually writing to the unit test DB? I'm assuming it is, but good to confirm. Either way, seems like creating these is doing some nontrivial work.)

I'm sure there must be ways to insert 20,000 items more efficiently...

Reply:
Or maybe we could make it a permanent part of the DB instead of having to recreate it each time?

Reply:
I have un-crossed out my previous comment (see above)! If we want to have tests on massive sections, one way to work around the long setup time is to do them as an integration test rather than a unit test.

We could, as winter said, have a dedicated test DB for these sections. When the test starts, it can check for the existance of these resources - if theyre there, it doesnt create any resources, but if not it creates them. Im not sure if our test infrastructure is set up to handle this kind of end to end test at this time.

Reply:
@mvkski I'm not an expert here but I **think** Rails unit tests already use an actual database by default. Would be good to confirm, though.